build/openwisp_base/Dockerfile

```diff
@@ -1,65 +1,77 @@
-RUN pip install --prefix='/install' --upgrade ${OPENWISP_FIRMWARE_SOURCE}
+RUN pip install --user --upgrade ${OPENWISP_FIRMWARE_SOURCE}
 ARG OPENWISP_TOPOLOGY_SOURCE=https://github.com/openwisp/openwisp-network-topology/tarball/docker-openwisp-usage
-RUN pip install --prefix='/install' --upgrade ${OPENWISP_TOPOLOGY_SOURCE}
+RUN pip install --user --upgrade ${OPENWISP_TOPOLOGY_SOURCE}
 ARG OPENWISP_RADIUS_SOURCE=https://github.com/openwisp/openwisp-radius/tarball/docker-openwisp-usage
-RUN pip install --prefix='/install' --upgrade ${OPENWISP_RADIUS_SOURCE}
+RUN pip install --user --upgrade ${OPENWISP_RADIUS_SOURCE}
 ARG OPENWISP_IPAM_SOURCE=https://github.com/openwisp/openwisp-ipam/tarball/docker-openwisp-usage
-RUN pip install --prefix='/install' --upgrade ${OPENWISP_IPAM_SOURCE}
+RUN pip install --user --upgrade ${OPENWISP_IPAM_SOURCE}
 
 # here we try to install custom versions of the modules only if the
 # supplied argument does not equal the default value, because
 # otherwise these modules will have already been installed above
 ARG OPENWISP_CONTROLLER_SOURCE=default
 RUN if [ "$OPENWISP_CONTROLLER_SOURCE" != "default" ] ; then \
-    pip install --prefix='/install' --upgrade ${OPENWISP_CONTROLLER_SOURCE}; \
+    pip install --user --upgrade ${OPENWISP_CONTROLLER_SOURCE}; \
     fi
 ARG OPENWISP_NOTIFICATION_SOURCE=default
 RUN if [ "$OPENWISP_NOTIFICATION_SOURCE" != "default" ] ; then \
-    pip install --prefix='/install' --upgrade ${OPENWISP_NOTIFICATION_SOURCE}; \
+    pip install --user --upgrade ${OPENWISP_NOTIFICATION_SOURCE}; \
     fi
 ARG OPENWISP_USERS_SOURCE=default
 RUN if [ "$OPENWISP_USERS_SOURCE" != "default" ] ; then \
-    pip install --prefix='/install' --upgrade --force-reinstall ${OPENWISP_USERS_SOURCE}; \
+    pip install --user --upgrade --force-reinstall ${OPENWISP_USERS_SOURCE}; \
     fi
 ARG OPENWISP_UTILS_SOURCE=default
 RUN if [ "$OPENWISP_UTILS_SOURCE" != "default" ] ; then \
-    pip install --prefix='/install' --upgrade --force-reinstall ${OPENWISP_UTILS_SOURCE}; \
+    pip install --user --upgrade --force-reinstall ${OPENWISP_UTILS_SOURCE}; \
     fi
 ARG DJANGO_X509_SOURCE=default
 RUN if [ "$DJANGO_X509_SOURCE" != "default" ]; then \
-    pip install --prefix='/install' --upgrade --force-reinstall ${DJANGO_X509_SOURCE}; \
+    pip install --user --upgrade --force-reinstall ${DJANGO_X509_SOURCE}; \
     fi
 
 COPY ./openwisp_base/requirements.txt /tmp/openwisp-deploy-requirements.txt
-RUN pip install --prefix='/install' -r /tmp/openwisp-deploy-requirements.txt
+RUN pip install --user --upgrade -r /tmp/openwisp-deploy-requirements.txt
 ARG DJANGO_SOURCE=django~=4.0.0
-RUN pip install --prefix='/install' --upgrade ${DJANGO_SOURCE}
+RUN pip install --user --upgrade ${DJANGO_SOURCE}
 
 FROM SYSTEM
 
-COPY --from=PYTHON /install /usr/local
-COPY ./common/ /opt/openwisp/
+COPY --from=PYTHON --chown=openwisp:root /home/openwisp/.local/ /usr/local
```

Comment:
Might be worth checking if it's increasing the size of the container by a lot, but otherwise, GG! 😄

Reply:
These are the current sizes:

```
CONTAINER ID   IMAGE                                 COMMAND                  CREATED          STATUS          PORTS                                                                      NAMES                          SIZE
ceeddce1e1a9   openwisp/openwisp-dashboard:latest    "bash init_command.sh"   17 minutes ago   Up 17 minutes   8000/tcp                                                                   docker-openwisp_celery_1       35.9kB (virtual 994MB)
bbbdee687a12   openwisp/openwisp-api:latest          "bash init_command.sh"   17 minutes ago   Up 17 minutes   8001/tcp                                                                   docker-openwisp_api_1          33.9kB (virtual 994MB)
82af8c5f55e6   openwisp/openwisp-dashboard:latest    "bash init_command.sh"   17 minutes ago   Up 17 minutes   8000/tcp                                                                   docker-openwisp_celerybeat_1   52.3kB (virtual 994MB)
84bb4f818caf   openwisp/openwisp-websocket:latest    "supervisord --nodae…"   17 minutes ago   Up 17 minutes   8002/tcp                                                                   docker-openwisp_websocket_1    12.9kB (virtual 994MB)
3ca80bba5f70   openwisp/openwisp-dashboard:latest    "bash init_command.sh"   18 minutes ago   Up 17 minutes   8000/tcp                                                                   docker-openwisp_dashboard_1    38.2kB (virtual 994MB)
45042accb8d1   openwisp/openwisp-freeradius:latest   "/docker-entrypoint.…"   29 minutes ago   Up 29 minutes   0.0.0.0:1812-1813->1812-1813/udp, :::1812-1813->1812-1813/udp              docker-openwisp_freeradius_1   2.81kB (virtual 317MB)
7187641daf6a   openwisp/openwisp-nginx:latest        "/docker-entrypoint.…"   54 minutes ago   Up 54 minutes   0.0.0.0:80->80/tcp, :::80->80/tcp, 0.0.0.0:443->443/tcp, :::443->443/tcp   docker-openwisp_nginx_1        6.98kB (virtual 101MB)
cf8c5e6b576a   openwisp/openwisp-openvpn:latest      "sh init_command.sh"     54 minutes ago   Up 52 minutes   1194/tcp, 0.0.0.0:1194->1194/udp, :::1194->1194/udp                        docker-openwisp_openvpn_1      45.7kB (virtual 70.5MB)
886fc2e83c40   redis:alpine                          "docker-entrypoint.s…"   54 minutes ago   Up 54 minutes   6379/tcp                                                                   docker-openwisp_redis_1        0B (virtual 32.4MB)
01d543620f10   mdillon/postgis:11-alpine             "docker-entrypoint.s…"   54 minutes ago   Up 54 minutes   5432/tcp                                                                   docker-openwisp_postgres_1     63B (virtual 146MB)
0d5784cee9fe   influxdb:1.8-alpine                   "/entrypoint.sh infl…"   54 minutes ago   Up 54 minutes   8086/tcp                                                                   docker-openwisp_influxdb_1     0B (virtual 163MB)
803080d3fad2   openwisp/openwisp-postfix:latest      "sh init_command.sh"     54 minutes ago   Up 54 minutes   25/tcp                                                                     docker-openwisp_postfix_1      38.1kB (virtual 56.4MB)
```